### PR TITLE
Optional Fee Payer

### PR DESCRIPTION
**Pull request by 0xjinn**
### Description

1. Added "hasSponsor" field to all `InputGenerateTransactionData`. If set to `true`, feepayer address will be overrided to "0x0".
2. Fee payer will have to use "signTransactionAsPayer" to sign the transaction.

### Test Plan

```
Test Suites: 38 passed, 38 total
Tests:       1 skipped, 450 passed, 451 total
Snapshots:   0 total
Time:        24.384 s, estimated 25 s
```

### CODE REVIEW

**Comment by xbtmatt** [comment]
I love `hasSponsor` as a concept, but in the interest of keeping things consistent I think we should probably do `withFeePayer` or `hasFeePayer`

Although I'm partial to `withFeePayer` because `hasFeePayer` suggests it already has a fee payer, and it technically doesn't yet.

The clearest would be `willHaveFeePayer` but maybe that's too wordy

**Comment by xbtmatt** [comment]
Also another consistency thing:

`signTransactionAsFeePayer` or even `signAsFeePayer` (since there's no other thing you'll sign as a fee payer)

**Comment by 0xjinn** [comment]
@xbtmatt 

> I love `hasSponsor` as a concept, but in the interest of keeping things consistent I think we should probably do `withFeePayer` or `hasFeePayer`
> 
> Although I'm partial to `withFeePayer` because `hasFeePayer` suggests it already has a fee payer, and it technically doesn't yet.
> 
> The clearest would be `willHaveFeePayer` but maybe that's too wordy
> 
> Personally I'm a fan of wordy names since it's an sdk and it generally makes the devex better as long as it isn't an eyesore to look at (camelCase kinda sucks lol)

Yeah good call. I used `hasSponsor` mainly because of the concept "Sponsored Transaction". TBH I wonder is the official name, is it "Fee Payer Transaction", or "Sponsor Transaction"? Maybe @davidiw would know :)

File: src/api/transactionSubmission.ts [hunk 1]
```diff
@@ -110,6 +110,32 @@ export class TransactionSubmission {
...
+   *  rawTransaction: Uint8Array,
+   *  secondarySignerAddresses? : Array<AccountAddress>,
+   *  feePayerAddress?: AccountAddress
+   * }
+   * ```
+   *
+   * @return The signer AccountAuthenticator
+   */
+  /* eslint-disable class-methods-use-this */
+  signTransactionAsPayer(args: { signer: Account; transaction: AnyRawTransaction }): AccountAuthenticator {
```
Comment (0xmaayan) [inline src/api/transactionSubmission.ts:130]: I feel this whole function can simply be a boolean arguments

```
signTransaction(args: { signer: Account; transaction: AnyRawTransaction, asSponsor?:true }): AccountAuthenticator {
if(asSponsor){
 transaction.feePayerAddress = signer.accountAddress;
}
    return signTransaction({ ...args });
  }
```

File: src/transactions/types.ts [hunk 2]
```diff
@@ -267,7 +267,7 @@ export type InputSimulateTransactionOptions = {
  */
 export interface InputGenerateSingleSignerRawTransactionData {
   sender: AccountAddressInput;
-  feePayerAddress?: undefined;
+  hasSponsor?: undefined | false;
```
Comment (0xmaayan) [inline src/transactions/types.ts:270]: not sure you need the `false`? the intention here is that `hasSponsor` is always undefined right? then I think we shouldn't expect/let them set it to false

**Review by 0xmaayan — CHANGES_REQUESTED** [review]
boolean argument instead of a new function

**Comment by xbtmatt** [comment]
@0xjinn a bit ago we decided to go with fee payer in code and sponsored in non-code contexts, because changing the field names was backwards incompatibility already for some indexer stuff 😅

File: src/transactions/types.ts (hunk 2 above)
Comment (0xjinn) [reply src/transactions/types.ts:270]: Hm but if user decided to make it clear, and set the `hasSponsor` to false. I would expect that should also be a viable option.

File: src/api/transactionSubmission.ts (hunk 1 above)
Comment (0xjinn) [reply src/api/transactionSubmission.ts:130]: Yeah I was thinking about this vs a new function. I wonder what everyone's preference are. We can sync up real quick on monday standup.

File: src/transactions/types.ts (hunk 2 above)
Comment (0xmaayan) [reply src/transactions/types.ts:270]: why would we want user to specify it if it doesnt make sense in this context as it is always undefined? even if they specify false.

Comment (0xjinn) [reply src/transactions/types.ts:270]: Okay I think I see what you mean. Since the Fee payer transaction data is marked as specifically, `hasSponsor: true`, we dont need to let user to specify the false, and its always undefine for other transaction data. 

Will update.

File: tests/e2e/transaction/helper.ts [hunk 3]
```diff
@@ -116,15 +116,17 @@ export const rawTransactionMultiAgentHelper = async (
   secondarySignerAccounts: Array<Account>,
   feePayerAccount?: Account,
```
Comment (0xmaayan) [inline tests/e2e/transaction/helper.ts:117]: should we change it to be a boolean flag?

**Review by 0xmaayan — APPROVED** [review]
lgtm. lets also update CHANGELOG

File: tests/e2e/transaction/helper.ts (hunk 3 above)
Comment (0xjinn) [reply tests/e2e/transaction/helper.ts:117]: Nah lets not do that, we still need the feed payer account to actually sign the transaction.
